### ISSUE TITLE
Sleigh: allow GOTO VARSYM

### ISSUE THREAD

**Issue by yatli** [NationalSecurityAgency/ghidra#2019]
Currently GOTO only allows OPERANDSYM, which requires the symbol to be an operand in the constructor.

It's not allowed to use a variable node as the GOTO destination.
That is, the following example is not possible:

```
:bra rn is rn & inst=... {
    local dest:3 = rn + inst_start;
    goto dest;
}
```

However it can be worked around by sending `dest` into a macro. The macro input parameter is recognized as an operand in the yacc file.

The export also works, but not as clean as writing the GOTO statement directly.

**Comment by mumbel** [comment]
You should probably read the docs again and maybe reference some other processors.  `goto [dest];` btw

**Comment by yatli** [comment]
Thanks for the tip!

I was looking at the example in the doc:

```
:ret is opcode=8 {
  sp=sp+4;
  tmp:4 = * sp;
  return [tmp];
}
```

... and thinking BRANCHIND has an extra dereference like a pointer to an address is stored on stack and it's returning to that.

So that's not the case, right?

**Comment by mumbel** [comment]
This is my understanding of it at least... think more of it as 'is this a constant': `goto inst_next`or `goto imm32` or `goto imm32+inst_next` or `goto <internal_label>`... Those are all know.  Add in a register which could be anything and ghidra will need to compute it, so it needs `[]`

**Comment by yatli** [comment]
@mumbel interestingly, this works:

```sleigh
:^inst is inst & loopActive=1 & loopEnd=inst_start [loopActive=0; globalset(inst_next, loopActive);] { 
  build inst;
  lc = lc - 1;
  local tmp:3=0;
  stack_pop(tmp, sr);
  stack_pop(la, lc);
  if (lc != 0) goto <post_go>; 
  goto [tmp];
  <post_go>
}
```

but this doesn't:

```sleigh
:^inst is inst & loopActive=1 & loopEnd=inst_start [loopActive=0; globalset(inst_next, loopActive);] { 
  build inst;
  lc = lc - 1;
  local tmp:3=0;
  stack_pop(tmp, sr);
  stack_pop(la, lc);
  if (lc != 0) goto [tmp]; 
}
```

**Comment by mumbel** [comment]
Hrmm, any luck with something like this instead?
```
if (lc == 0) goto inst_next;
goto [tmp];
```

**Comment by yatli** [comment]
That also works. 
Looks like the problem is specific to the `if (expr) goto [dest]` form, giving "invalid dynamic target used in goto destination"

**Comment by yatli** [comment]
btw a quick question:

this is the primary rule in the instruction table to replicate the hardware do-loop behavior.

```
:^inst is inst & loopActive=1 & loopEnd=inst_start [loopActive=0; globalset(inst_next, loopActive);] { 
  build inst;
  lc = lc - 1;
  local tmp:3=ssh;
  if (lc == 0) goto <end_of_loop>; 
  goto [tmp];
  <end_of_loop>
  stack_pop(tmp, sr);
  stack_pop(la, lc);
}
```

In the pattern section I have `loopEnd=inst_start`, where `loopEnd` is a context variable set by a `do` instruction, and I compare it to the current instruction address to trigger the loop behavior. But it looks like this pattern is never hit because `inst_start` evaluates to 0 -- found this with the ghidra sleigh debugging script:

```
check pattern[1 of 2]: {line# 1729} <inst>
   context pattern: mask=11111111.11111111.11111111.10000000
          context(0..31)=00000000.00000001.01011100.10000000
             match-value=00000000.00000000.00000000.10000000 Failed
               loopActive(24,24) == 0x1 Match
               loopEnd(0,23) == 0x0 Failed (=0x15c)
check pattern[2 of 2]: {line# 1740} <inst>
   byte pattern: always-Matched
```

pattern 1 is the loop behavior, and pattern 2 is the fallback default (just build the instruction).

**Comment by shuffle2** [comment]
I was also trying to implement loop instructions for an architecture and came across this problem / github issue. In my case it was further complicated by the fact that the desired jump target was held in a register in a `ram_space`. I found it impossible to get ghidra to generate nice pcode/decompilation unless I just changed the registers to be in `register_space`. Just a note in case someone else comes across this...
